# Look at the file ripple-transactions-jan-2013-dec-2016

In [1]:
import pandas as pd

Let's import the data

In [2]:
data = pd.read_csv('../data/transactions/original/ripple-transactions-jan-2013-dec-2016.txt', header = None, names = ["hash", "sender", "receiver", "currency","amount1","amount2","ledger","sourceTag","DestinationTag","9","10"])

In [3]:
data.head()

hash  \
0  5C0504B229CD8EE803B01D78868B8F4B595D8B1D03636B...   
1  56022C64D29974C116591563DC8A80547907BE38EEB134...   
2  CCA2EB48BC106C7F4692BF71819EC236C9E09E37BCF38C...   
3  423708E30D94335FE2858C11B3D57390669A20FCE6C485...   
4  C1E342D3168740197C7E6B84110CEF9A338EA0D0EF5697...   

                               sender                            receiver  \
0  rfe8yiZUymRPx35BEwGjhfkaLmgNsTytxT  rsdbFqwyGYX5fSvAmMEXBH2hRyBT8QfcUS   
1  rfe8yiZUymRPx35BEwGjhfkaLmgNsTytxT  rsdbFqwyGYX5fSvAmMEXBH2hRyBT8QfcUS   
2  rsdbFqwyGYX5fSvAmMEXBH2hRyBT8QfcUS  rhjyoR8rpLMKM47pUgTjdNuVDJ1CrYYBwT   
3  rMTzGg7nPPEMJthjgEBfiPZGoAM7MEVa1r  rsdbFqwyGYX5fSvAmMEXBH2hRyBT8QfcUS   
4  rLz6TEYTsSZivwgXssq8ZSmTsBSFTMCw7h  rffqWVgnv9bPKaWHHVUDMjdAEgab3LGjZ1   

  currency  amount1  amount2   ledger  sourceTag  DestinationTag  9  \
0      XRP      0.0    100.0  7227428        NaN             NaN  1   
1      XRP      0.0    100.0  7224947        NaN             NaN  1   
2      XRP   2000.0      0.0  5541569        NaN             NaN  1   
3      XRP   2020.0      0.0  3333190        NaN             NaN  1   
4      XRP    100.0      0.0  6965416        NaN             NaN  1   

           10  
0  1402920660  
1  1402908860  
2  1394924250  
3  1384463290  
4  1401681860

By inspecting various fields and compring them in the data found on the https://xrpcharts.ripple.com/#/ it seems that
 - the first column is the hash
 - the second one is the sender address, 
 - third is receiver
 - fourth is currency
 - fifth and six are the amount of the transaction (since the data was in csv format and amount were given with a "," delimiter we need to combine fifth and 6 columns.
 - six is the ledger number
 - the columns seven and eight are respectively sourceTag and DestinationTag
 - 9 and 10 are not obvious at first look (And i don't know if it is a CSV error as in column 5 and 6)

Let's try to combine amount1 and amount2. The data in these two columns is in float format.<br>
We will first convert them to integers then to string and combine them with a "." and convert back to float.

In [4]:
print(data.amount1.isna().unique())
print(data.amount2.isna().unique())

[False  True]
[False  True]


It seems we have NA values in amount1 and amount2

In [5]:
data[data.amount1.isna()]

hash sender receiver  \
24053310  E1709FAA24AEB0A9FE4B4471BF59A55052A33A1085604B...    NaN      NaN   

         currency  amount1  amount2    ledger  sourceTag  DestinationTag  9  \
24053310      NaN      NaN      NaN  23670686        NaN             NaN  2   

                  10  
24053310  1472360631

In [6]:
data[data.amount2.isna()]

hash sender receiver  \
24053310  E1709FAA24AEB0A9FE4B4471BF59A55052A33A1085604B...    NaN      NaN   

         currency  amount1  amount2    ledger  sourceTag  DestinationTag  9  \
24053310      NaN      NaN      NaN  23670686        NaN             NaN  2   

                  10  
24053310  1472360631

Indeed at line 24053310 the data seems corrupted. Lets drop this line.

In [7]:
data.dropna(axis=0, how='any', subset= ["amount1","amount2"], inplace=True)

Now we can convert to int 

In [8]:
data.amount1 = data.amount1.astype('int64')
data.amount2 = data.amount2.astype('int64')

And combine amount1 and amount2 in a new column called amount.

In [9]:
data["amount"] = data.amount1.astype(str).str.cat(data.amount2.astype(str), sep='.')
data.amount = data.amount.astype('float64')
data.drop(["amount1", "amount2"], axis = 1) #remove old column
data = data[['hash','sender','receiver','currency','amount','ledger','sourceTag','DestinationTag','9','10']] #re-order columns

In [10]:
data

hash  \
0         5C0504B229CD8EE803B01D78868B8F4B595D8B1D03636B...   
1         56022C64D29974C116591563DC8A80547907BE38EEB134...   
2         CCA2EB48BC106C7F4692BF71819EC236C9E09E37BCF38C...   
3         423708E30D94335FE2858C11B3D57390669A20FCE6C485...   
4         C1E342D3168740197C7E6B84110CEF9A338EA0D0EF5697...   
5         F202202922C5D616096A51E76D0C65895620DEAB9517C3...   
6         5234065EB0115DFE0D1A71D7C2640EE798254A25E302D0...   
7         358E5F076F3276C0516C271E1987BE619CB84ABF352B87...   
8         8ADB8D496F9F21799919F23FCF1B84A2A1EDCBA884B61A...   
9         40655777FE754D15872BB77943811C8596239BCBC21012...   
10        176FCFC0A4607B82DDC0847BA0C101858296238226F94D...   
11        39C34F102D354D34326EF13594F1E00938E23BE294F79A...   
12        92DE6B0A1F01BD144E51724C3000253F380540DF670010...   
13        8E3B1173F5B3ACA9CF299169777E55A27212D707B9281D...   
14        4F328B5E4A6A3C0D8A549D1A67F440E9DCF15B5224A1BB...   
15        04E3E0D6D80413B07629B341C2ECFCCD91612DFBFC63B6...   
16        0B4EF7A42CE1258C2762EE926C18F0BD03D61654E11844...   
17        A7078F84527B6526A8ED6566C5A7A1B087BF18DC77B4BC...   
18        DAAE5F7CE4BAB832B2FAE1071018C8E4C65F40DED6ACEA...   
19        6C41964077A84DA258F48898E1DA92ED9A42141C6D9F18...   
20        2B67C4E0F52697EBF1CEEEC8392FCC619E1D978E9D7C66...   
21        BACAF6E8A75EE3E353395D546EE3958127CCD736254E03...   
22        D595D90F6F59C1D4ADCBDD8D79FC540F1C8213BD273A02...   
23        C2E2C762948C05FEC1040255F2C4C30D1D5C67A6926BF1...   
24        0253976B33FA88EF68CA2676245269A046FA05EA2B2351...   
25        A06B109594D45E25827732FC7C82139B50CEBF2AAC036E...   
26        475B82507CF376604DFA141A2032417F3873C2FA45845E...   
27        7DCD8673748908D554152E8CA615091349E56E3B6A6636...   
28        7143C5D37D1923F24CA9D891C680EDBD83DA6F411AA54D...   
29        E19951A7D504F978CC66DD0068DF694F0872F43FF3F7F0...   
...                                                     ...   
27406846  6935E30B2162F4E749E0FF299AD639FDD7C98F0CD481BB...   
27406847  8CC9D5F7650498F48E2BDBC27A175B2ABF2C3537DB3441...   
27406848  7132207F682CA9939DEAA1B5AFD6F83F52884AD85BA190...   
27406849  17CAAB601A7E31F830DABE28D1F010BC1C99BFB15E4945...   
27406850  5495F511C090EF2BB30DC3DF81064A53FDFF9200E9AD50...   
27406851  AAC0183F73E898A72472F6D4B686471C3215CD186FEB57...   
27406852  CED1ED66B2D389B126073F1B9408CCB25830DD754D075F...   
27406853  E6EFD82E7F6E3467B056CF99FBC023252D32815E23A0CB...   
27406854  D56C3EEDD524E25014558BB467151DEF46FCDC1930BE40...   
27406855  C46F0AEA2B90442CD04A704673D4A67CBF1DC61600FD48...   
27406856  DFD259F9B1B848FD30F3D2411A2FF70D8BDF666EF6C446...   
27406857  BC61D7CF7D3FAA9B42AF92F778EB3FAC001BDBA90F607B...   
27406858  0464D743C1A6F00FE1D4301448B3FA77DB755D16BC5AD3...   
27406859  755A1BF20B2D4A4A8FE73FC78EE9A6F0FC5868570243F9...   
27406860  5B4F418623F4EEE2CAE15B8B7AEFC3F106271BB5DD5AB3...   
27406861  A718E7A9FF023C09C8F5F0A1579735DAE47517E2C6FE63...   
27406862  AE0E03985AC2E3B947BF84581EF9AB4D674FC60FB131B1...   
27406863  EB246AA7C5F9714C627FE421C04B16A48848732B10FC90...   
27406864  36D947EA1722005EBDA7552B881CAB5572D5D8C6D29809...   
27406865  B6A8A613C30500864278F13C08C5AD36BEECA18595C665...   
27406866  BFEFDFB512B1C06D3256A4E0FCFDEDAE96F374373291B7...   
27406867  058B0CF11988ECF76AA47781DDB65BCAD531450E473B8A...   
27406868  C408283D525A6B414A17FAC62B0AF31D9495117290952D...   
27406869  49F05B2CB8B1EF8111F3FB9740CF5BFF067399FDE6EF38...   
27406870  63962512496115D176C514EBC26565087BF0D14AF25B80...   
27406871  7BD5A39F2590857412441D2BA73530AD0270AD8314453E...   
27406872  732AF2ED57E709B8FFEFEE25C7833837AD7EED578FEC1C...   
27406873  D942FB4E0106E6D907C50A93E132833A60B6C3965B68E9...   
27406874  4D4137F4CC3F66E5D3A09A0DDCE1E37F5A619C0BB55AC2...   
27406875  FF3884E622BFB41EEB863EF5E507D6B08BFF69C6CE46C0...   

                                      sender  \
0         rfe8yiZUymRPx35BEwGjhfkaLmgNsTytxT   
1         rfe8yiZUymRPx35BEwGjhfkaLmgNsTytxT   
2   

In [11]:
data.to_csv('../data/transactions/ripple-transactions-jan-2013-dec-2016.csv')